### PR TITLE
8283573: [lworld] Make Object an abstract class

### PR DESCRIPTION
**Pull request by fparain**
---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8283573](https://bugs.openjdk.java.net/browse/JDK-8283573): [lworld] Make Object an abstract class

### Reviewers
 * [David Simms](https://openjdk.java.net/census#dsimms) (@MrSimms - Committer)
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/valhalla pull/602/head:pull/602` \
`$ git checkout pull/602`

Update a local copy of the PR: \
`$ git checkout pull/602` \
`$ git pull https://git.openjdk.java.net/valhalla pull/602/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 602`

View PR using the GUI difftool: \
`$ git pr show -t 602`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/valhalla/pull/602.diff">https://git.openjdk.java.net/valhalla/pull/602.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/c1/c1_GraphBuilder.cpp [hunk 1]
```diff
@@ -2537,7 +2537,11 @@ void GraphBuilder::new_instance(int klass_index) {
   bool will_link;
   ciKlass* klass = stream()->get_klass(will_link);
   assert(klass->is_instance_klass(), "must be an instance klass");
-  NewInstance* new_instance = new NewInstance(klass->as_instance_klass(), state_before, stream()->is_unresolved_klass());
+  ciKlass* ik = klass;
+  if (ik == compilation()->env()->Object_klass() && compilation()->env()->Object_klass()->is_abstract()) {
```
Comment (MrSimms) [inline src/hotspot/share/c1/c1_GraphBuilder.cpp:2541]: Does adding if EnableValhalla as the first condition, or a "// Valhalla" comment, make this clearer ? There are a few places in the change set with the same review comment, but I won't repeat.

**Review by MrSimms — COMMENTED** [review]
The pattern:

`if (ik == vmClasses::Object_klass() && vmClasses::Object_klass()->is_abstract()) {
  ik = vmClasses::Identity_klass();
}`

Keep repeating itself, could this be an inline method somewhere, maybe `valhalla_check_replace_object_identity()` (i.e. self documenting)

**Comment by fparain** [comment]
> The pattern:
> 
> `if (ik == vmClasses::Object_klass() && vmClasses::Object_klass()->is_abstract()) { ik = vmClasses::Identity_klass(); }`
> 
> Keep repeating itself, could this be an inline method somewhere, maybe `valhalla_check_replace_object_identity()` (i.e. self documenting)

A single method would not be possible, because some tests are using Klass* while others are using ciKlass*.

**Comment by MrSimms** [comment]
> A single method would not be possible, because some tests are using Klass* while others are using ciKlass*.

Yeah it looks only 2 with ciKlass, hardly worth it

File: src/java.base/share/classes/java/lang/Identity.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2022, 2022, Oracle and/or its affiliates. All rights reserved.
```
Comment (TobiHartmann) [inline src/java.base/share/classes/java/lang/Identity.java:2]:
```suggestion
 * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
```

Comment (TobiHartmann) [reply src/java.base/share/classes/java/lang/Identity.java:2]: (This has been discussed in mainline before. If both years are equal, there should only be one)

**Review by TobiHartmann — APPROVED** [review]
Compiler changes look good to me.

**Comment by fparain** [comment]
Last update addresses the copyright format issue. It also removes the EnableValhalla condition in the interpreter changes to allow the VM to be fully functional even with -XX:-EnableValhalla (we need this configuration to track regressions with mainline, and the current JDK sources support only one version of java.lang.Object which is now abstract in the lworld branch). Last change is the addition of the aarch64 support.

**Review by MrSimms — APPROVED** [review]
Looks good

**Review by TobiHartmann — APPROVED** [review]
Looks good.

**Pull request closed without merging**